VFB-419: Common Delivery Instructions

alexandrutrifu-urzica opened the pull request:
## What's changed
Client Delivery Instructions can now be edited from the _Edit Parcel_ Form.

## Screenshots / Videos
| Before     | After      |
|------------|------------|
| <img width="899" height="300" alt="image" src="https://github.com/user-attachments/assets/ca4c385a-64ad-4084-9c47-77946cc64e8e" />| <img width="884" height="491" alt="image" src="https://github.com/user-attachments/assets/be193e86-b6d3-4a03-988b-4cd7454a836a" />|

## Checklist
- [x] The ticket is up-to-date - Please document any deviations from the original approach if there is any.
- [x] I have documented the testing steps for QA
- [x] I have self-reviewed this PR
- [x] Make sure you've verified it works via `npm run dev`
- [x] Make sure you've verified it works via `npm run build` and `npm run start`
- [x] Make sure you've fixed all linting problems with `npm run lint_fix`
- [x] Make sure you've tested via `npm run test`

alexandrutrifu-urzica committed:
VFB-419: fetched client delivery instructions in fetchParcel() method

# Conflicts:
#	src/common/fetch.ts

alexandrutrifu-urzica committed:
VFB-419: added delivery instructions field

# Conflicts:
#	src/app/parcels/form/ParcelForm.tsx

alexandrutrifu-urzica committed:
VFB-419: added parcel-specific card

anamariasuditu committed:
VFB-419: updated client delivery instructions through parcel form

anamariasuditu committed:
VFB-419: added update query

anamariasuditu committed:
VFB-419: added import comma

LeonardCiobanu-Softwire commented on src/app/parcels/form/formSections/DeliveryInstructionsCard.tsx:
```diff
@@ -0,0 +1,40 @@
+} from "@/components/Form/formFunctions";
+import GenericFormCard from "@/components/Form/GenericFormCard";
+
+export const MAX_CHARACTERS = 380;
```
Why do you need a limit for the number of characters?

alexandrutrifu-urzica replied:
The limit had already been imposed.

LeonardCiobanu-Softwire commented on src/app/parcels/form/submitFormHelpers.ts:
```diff
@@ -44,46 +45,67 @@ export const insertParcel: InsertParcel = async (parcelRecord) => {
-    return { parcelId: primaryKey, error: null };
-};
+        const { error: updateDeliveryInstructionsError } = await updateClientDeliveryInstructions(
+            parcelRecord.client_id == undefined ? "" : parcelRecord.client_id,
```
Why would the client_id be undefined? Maybe handle it differently than just send an empty string.

alexandrutrifu-urzica replied:
This is the way it was originally defined.

AlexandraIrinaMardale replied:
And what happens if we do end up having an undefined client id and we pass on the empty string?

alexandrutrifu-urzica replied:
The `UpdateSchema` generated for the _parcels_ table automatically labels the `client_id` field as optional, hence the possibility of it being passed forward as `undefined`.

This `client_id` is used to identify the client whose instructions should be modified, so either it being null or an empty string would result in the same error, which is then logged below.

I've noticed the associated `parcelRecord` object had already defined the client_id field as `clientId || fields.clientId || ""`, with the empty string alternative included, so we decided to keep it this way. Again, the behaviour would remain the same if we opted for passing it on as null.

AlexandraIrinaMardale replied:
Ok, but when you pass the empty string, that results in a bad request. Can we check for empty before we send the request please?

alexandrutrifu-urzica replied:
I added some checks to both the insert and update methods.

AlexandraIrinaMardale commented on src/app/clients/form/formSections/DeliveryInstructionsCard.tsx:
Can you revert the changes in this file please?

alexandrutrifu-urzica replied:
I removed this client-specific card and kept the one in the _/common/formSections_ folder as a common component.
When rebasing the branch, I'll make sure to drop this initial commit as well.

AlexandraIrinaMardale commented on src/app/parcels/form/formSections/DeliveryInstructionsCard.tsx:
Can you find a way to not duplicate this code, please?

alexandrutrifu-urzica replied:
The new common Card component uses a _CommonCardProps_ type, which seems to have solved the issue.

alexandrutrifu-urzica committed:
VFB-419: Client Form now uses common DeliveryInstructionsCard from parcels

alexandrutrifu-urzica committed:
VFB-419: removed client-specific DeliveryInstructionsCard

alexandrutrifu-urzica committed:
VFB-419: DeliveryInstructionsCard is now used in both client and parcel forms

alexandrutrifu-urzica committed:
VFB-419: defined common card props and common fields to be used in the DeliveryInstructionsCard

alexandrutrifu-urzica committed:
VFB-419: moved common DeliveryInstructionsCard to /common/formSections

alexandrutrifu-urzica committed:
VFB-419: fixed import errors

LeonardCiobanu-Softwire commented on src/app/parcels/form/ParcelForm.tsx:
```diff
@@ -69,6 +70,7 @@ export interface ParcelFields extends Fields {
     collectionSlot: string | null;
     collectionCentre: string | null;
     lastUpdated: string | undefined;
+    deliveryInstructions: string | null | undefined;
```
I believe type `string | null` would be enough.

alexandrutrifu-urzica replied:
We updated the type definition and added an extra _null check_ in the `ParcelForm` to remove the possibility of the field becoming undefined👍.

LeonardCiobanu-Softwire reviewed (commented):
👍

alexandrutrifu-urzica committed:
VFB-419: removed undefined type in deliveryInstructions definition

mihaicutaiulia commented on src/app/clients/form/ClientForm.tsx:
```diff
@@ -76,7 +76,7 @@ export interface ClientFields extends Fields {
     petFood: BooleanGroup;
     otherItems: BooleanGroup;
-    deliveryInstructions: string;
+    deliveryInstructions: string | null | undefined;
```
can we apply the same changes here as we did for the parcel form so that `undefined` is no longer needed?

anamariasuditu replied:
> can we apply the same changes here as we did for the parcel form so that `undefined` is no longer needed?

Yes, you are right, I applied the same changes as in the parcel form

mihaicutaiulia commented on src/app/clients/form/ClientForm.tsx:
```diff
@@ -64,7 +64,7 @@ export interface ClientFields extends Fields {
     children: Person[];
     numberOfChildren: number;
-    listType: ListType | null;
+    listType?: ListType | null;
```
in what scenarios do we need `listType?` instead of a required `listType` ?

anamariasuditu replied:
> in what scenarios do we need `listType?` instead of a required `listType` ?

i modified it to a required `listType`

mihaicutaiulia commented on src/app/parcels/form/ParcelForm.tsx:
```diff
@@ -17,46 +21,43 @@ import {
 export interface ParcelFields extends Fields {
     clientId: string | null;
-    listType?: ListType;
+    listType?: ListType | null;
```
same here

anamariasuditu replied:
> same here

i modified it to requiered `listType`

mihaicutaiulia commented on src/app/parcels/form/ParcelForm.tsx:
```diff
@@ -285,21 +297,25 @@ const ParcelForm: React.FC<ParcelFormProps> = ({
         const parcelRecord = {
             client_id: clientId || fields.clientId || "",
-            list_type: fields.listType,
+            list_type: fields.listType ?? undefined,
```
also here

anamariasuditu replied:
> also here

Because in `databaseTypeFIles`, list_type is declared like this: `list_type?: "regular" | "hotel" | undefined` so 
type "regular" | "hotel" | null is not assignable to type "regular" | "hotel" | undefined

anamariasuditu committed:
VFB-419: removed undefined from deliveryInstructions in ClientForm and listType

# Conflicts:
#	src/app/clients/form/ClientForm.tsx
#	src/app/parcels/form/ParcelForm.tsx

AlexandraIrinaMardale commented on src/app/clients/updateDeliveryInstructions.ts:
```diff
@@ -0,0 +1,34 @@
+        const logId = await logErrorReturnLogId("update client delivery instructions failed", {
+            error,
+        });
+        await sendAuditLog({ ...baseAuditLogProps, wasSuccess: false, logId });
```
I think you're sending the audit log twice? Here and in `submitFormHelper`

alexandrutrifu-urzica replied:
Correct, I've deleted this call and only kept the one in the `submitFormHelper`, thank you!

AlexandraIrinaMardale commented on src/app/parcels/form/ParcelForm.tsx:
```diff
@@ -17,46 +21,43 @@ import {
 export interface ParcelFields extends Fields {
     clientId: string | null;
-    listType?: ListType;
+    listType?: ListType | null;
```
Why did you need to make any changes to the list type? 🤔

alexandrutrifu-urzica replied:
There were some compatibility issues at first, but I fixed everything on a new rebase - now `listType` should be required and initialized with a default value.

alexandrutrifu-urzica replied:
**_Update_**: @AlexandraIrinaMardale I've since noticed that the additional null type is **indeed required** for our ticket. 

Currently, the `ClientFields` interface on `intern-dev` defines `listType` as `ListType | null`. To successfully refactor the `DeliveryInstructionsCard` to be common for both parcel and client forms, we needed an extra interface for the common `CardProps` it needed to receive. Therefore, the `ParcelFields` interface needed to be compatible with `ClientFields` - hence the refactoring of the `listType` parcel-field as possibly null as well.

AlexandraIrinaMardale commented on src/app/parcels/form/ParcelForm.tsx:
```diff
@@ -102,6 +109,7 @@ export const initialParcelFields: ParcelFields = {
     collectionSlot: null,
     collectionCentre: null,
     lastUpdated: undefined,
+    deliveryInstructions: null, // this should be the client's field
```
Can you remove this comment please?

AlexandraIrinaMardale commented on src/common/fetch.ts:
```diff
@@ -72,6 +75,10 @@ export const fetchParcel = async (
                 is_shown
+            ),
+            client: clients(
+                primary_key,
```
Do you need the primary key here?

alexandrutrifu-urzica replied:
I removed it, I had forgotten it wasn't necessary for identifying the client object, thank you!

AlexandraIrinaMardale reviewed (commented):
Left some comments 😄

alexandrutrifu-urzica commented:
**FIXED**: Another thing I noticed was that the `Delivery Instructions Card` in the `Add Parcel Form` has not been synchronized with the client's fields - this means that the Parcel Form field is initially empty.

I will look into this and come back with a fix.

alexandrutrifu-urzica committed:
VFB-419: refactored code

alexandrutrifu-urzica committed:
VFB-419: removed client's primary key from fetch query

alexandrutrifu-urzica committed:
VFB-419: removed duplicate sendAuditLog call

alexandrutrifu-urzica committed:
VFB-419: added option to edit delivery instructions to AddParcelForm

alexandrutrifu-urzica committed:
VFB-419: fixed linter error

alexandrutrifu-urzica committed:
VFB-419: added null type to listType definition for compatibility

AlexandraIrinaMardale commented on src/app/parcels/form/submitFormHelpers.ts:
```diff
@@ -40,50 +44,85 @@ export const insertParcel: InsertParcel = async (parcelRecord) => {
+            "Error with update: parcel data",
+            updateDeliveryInstructionsError
+        );
+        await sendAuditLog({ ...auditLog, wasSuccess: false, logId });
```
You need to update the audit log to also include the delivery instructions

AlexandraIrinaMardale commented on src/common/fetch.ts:
```diff
@@ -72,6 +75,9 @@ export const fetchParcel = async (
                 primary_key,
                 is_shown
+            ),
+            client: clients(
```
Nitcpick: can this be `client:clients (` and same for packing slot

AlexandraIrinaMardale commented on src/common/fetch.ts:
```diff
@@ -12,6 +12,8 @@ type CollectionCentre = Pick<
 
 type PackingSlot = Pick<Schema["packing_slots"], "primary_key" | "is_shown" | "name">;
 
+type Client = Pick<Schema["clients"], "delivery_instructions">;
```
If we are just picking the delivery instructions, the type should reflect that

alexandrutrifu-urzica replied:
We renamed it to `ClientWithDeliveryInstructions`, following the naming convention for Parcel types (e.g. `ParcelWithCollectionCentreAndPackingSlot`).

alexandrutrifu-urzica committed:
VFB-419: refactored type name

alexandrutrifu-urzica committed:
VFB-419: added delivery instructions to audit log content

alexandrutrifu-urzica committed:
VFB-419: fixed rebase issues

alexandrutrifu-urzica committed:
VFB-419: added parcel notes card back

alexandrutrifu-urzica committed:
VFB-419: added check for undefined client_id in updateParcel()

alexandrutrifu-urzica committed:
VFB-419: refactored error logging and added empty client id check in insertParcel()

AlexandraIrinaMardale commented on src/common/fetch.ts:
```diff
@@ -68,10 +72,13 @@ export const fetchParcel = async (
                 primary_key,
                 is_shown
+            ),
+            client:clients(
```
Can you add a space before the parenthesis? 😁

AlexandraIrinaMardale commented on src/app/parcels/form/submitFormHelpers.ts:
```diff
@@ -40,50 +44,110 @@ export const insertParcel: InsertParcel = async (parcelRecord) => {
-        .update(parcelRecord, { count: "exact" })
-        .eq("primary_key", primaryKey)
-        .eq("last_updated", parcelRecord.last_updated);
+export const updateParcel: UpdateParcelWithPrimaryKey =
```
We need this update to be atomic, i.e when we update the parcel, we need to make sure that either the entire update succeeds or fails. Currently, it might update the parcel details but fail on the delivery instructions or the other way around.

alexandrutrifu-urzica replied:
I added 2 new Supabase functions and now the parcel insert/update operations also handle the client's delivery instructions.

alexandrutrifu-urzica committed:
VFB-419: added migrations to create new insert/update functions

alexandrutrifu-urzica committed:
VFB-419: added missing space

alexandrutrifu-urzica committed:
VFB-419: refactored insert/update parcel methods to use new supabase functions

alexandrutrifu-urzica committed:
VFB-419: deleted unused updateDeliveryInstructions method

DanielaNegrut commented on src/app/clients/form/ClientForm.tsx:
```diff
@@ -78,7 +78,7 @@ export interface ClientFields extends Fields {
     petFood: BooleanGroup;
     otherItems: BooleanGroup;
-    deliveryInstructions: string;
+    deliveryInstructions: string | null;
```
Why is this allowed a `null` value now?

anamariasuditu replied:
> Why is this allowed a `null` value now?

Because `deliveryInstructions` can be `null`, if another value is not provided

DanielaNegrut commented on src/app/parcels/form/ParcelForm.tsx:
```diff
@@ -32,29 +51,13 @@ import {
 export interface ParcelFields extends Fields {
     clientId: string | null;
-    listType?: ListType;
+    listType: ListType | null;
```
Why was this changed from `undefined` to `null`?

anamariasuditu replied:
> Why was this changed from `undefined` to `null`?

Because we now changed the default value to `Regular`, so it's no longer undefined by default

DanielaNegrut replied:
But it sill is null by default?

AlexandraIrinaMardale replied:
@anamariasuditu what was the conclusion on this?

anamariasuditu replied:
> @anamariasuditu what was the conclusion on this?

`ListType` is `Regular` by default, and the `listType` field can be either of `ListType` type, or `null`, making the field optional. The `| null` makes the field nullable, while the default value in `initialParcelFields `ensures it always has a sensible starting value.

DanielaNegrut commented on src/app/parcels/form/ParcelForm.tsx:
```diff
@@ -177,7 +191,7 @@ const databaseErrorMessageFromErrorType = (
     initialFields,
     initialFormErrors,
-    clientId,
+    clientId = null,
```
Why is this default null, is there any case where there is a parcel without a client id?

anamariasuditu replied:
> Why is this default null, is there any case where there is a parcel without a client id?

You are right, there shouldn't be a parcel without a client id. I updated it, thank you!

DanielaNegrut commented on src/app/parcels/form/ParcelForm.tsx:
```diff
@@ -79,15 +83,20 @@ export interface ParcelErrors extends FormErrors<ParcelFields> {
 export const initialParcelFields: ParcelFields = {
     clientId: null,
-    listType: undefined,
+    listType: "regular",
```
How come this has been initialized now and wasn't before?

anamariasuditu replied:
> How come this has been initialized now and wasn't before?

Because the `listType` can be `Regular `or `Hotel`, and it should be set on `Regular `by default in the form, unless the user changes it.

DanielaNegrut commented on src/app/parcels/form/submitFormHelpers.ts:
```diff
@@ -1,89 +1,117 @@
+    await sendAuditLog({
+        ...auditLog,
+        wasSuccess: true,
+        parcelId: parcelDataWithCount.parcel_primary_key || undefined,
```
In what case will the primary key be undefined?

alexandrutrifu-urzica replied:
There was an issue with the previous Supabase functions that resulted in the `primary key` being possibly null. I've since fixed this and the `undefined` type is no longer needed.

DanielaNegrut commented on src/app/parcels/form/submitFormHelpers.ts:
```diff
@@ -1,89 +1,117 @@
 
-type UpdateParcelErrorType = "failedToUpdateParcel" | "concurrentUpdateConflict";
+type UpdateParcelErrorType =
+    | "failedToUpdateParcel"
+    | "concurrentUpdateConflict"
+    | "failedToUpdateDeliveryInstructions";
```
Let's order the error types so the ones starting with `failed to update ...` are next to each other.

anamariasuditu replied:
> Let's order the error types so the ones starting with `failed to update ...` are next to each other.

Yes, I updated it

DanielaNegrut commented on src/common/fetch.ts:
```diff
@@ -37,6 +39,7 @@ export interface ParcelWithCollectionCentreAndPackingSlot {
     primary_key: string;
     voucher_number: string | null;
     last_updated: string | undefined;
+    client: ClientWithDeliveryInstructions | null;
```
The naming is a bit weird.

Having a whole field/property called `client` makes me think there is more information there, not just `delivery instructions`

anamariasuditu replied:
> The naming is a bit weird.
> 
> Having a whole field/property called `client` makes me think there is more information there, not just `delivery instructions`

We considered it would be more logic to name it client and have `client.delivery_instructions` when we access the field, instead of naming it `deliveryInstructions `and later call it `deliveryInstructions.delivery_instructions`

alexandrutrifu-urzica replied:
I renamed it to `clientWithDeliveryInstructions` to match the type for now. If we decide to reference more fields from the `clients` table later on, we can switch back to the old name.

DanielaNegrut commented on src/databaseTypesFile.ts:
```diff
@@ -982,10 +997,18 @@ export type Database = {
         time: string | null
         is_active: boolean | null
       }
+      insert_parcel_result: {
+        parcel_primary_key: string | null
+        rows_inserted: number | null
+      }
```
Does it make sense to have such a small difference in the insert and update interfaces? Are there other situations where this is made?

alexandrutrifu-urzica replied:
After refactoring the migrations, these composite types should no longer be there.

DanielaNegrut commented on supabase/migrations/20250812092433_add_update_parcel_with_delivery_instructions_function.sql:
```diff
@@ -0,0 +1,45 @@
+$$
+DECLARE
+    rows_updated INTEGER;
+    client_id UUID := (parcel_record->>'client_id')::UUID;
```
Aren't there cases where this can be null?

anamariasuditu replied:
> Aren't there cases where this can be null?

The` delivery_instructions` field belongs to `client`, so the `client_id` can't be null if we try to pass the field to the `parcel`

DanielaNegrut commented on supabase/migrations/20250812101258_add_insert_parcel_with_delivery_instructions_function.sql:
```diff
@@ -0,0 +1,56 @@
+) RETURNS insert_parcel_result AS
+$$
+DECLARE
+rows_inserted INTEGER;
```
This isn't aligned the same as the other lines in the variable definition of the declare block

anamariasuditu replied:
> This isn't aligned the same as the other lines in the variable definition of the declare block

I updated it, thank you!

stefanghenescu-work commented on supabase/migrations/20250812092433_add_update_parcel_with_delivery_instructions_function.sql:
SQL syntax should be written in lower case. (e.g. `create type`).

anamariasuditu replied:
> SQL syntax should be written in lower case. (e.g. `create type`).

From what I've seen, most migrations are written in upper case.

stefanghenescu-work commented on supabase/migrations/20250812101258_add_insert_parcel_with_delivery_instructions_function.sql:
Same for lower case here

anamariasuditu committed:
VFB-419: refactored errors order

anamariasuditu committed:
VFB-419: refactored code

alexandrutrifu-urzica committed:
VFB-419: refactored clientWithDeliveryInstructions

alexandrutrifu-urzica committed:
VFB-419: updated supabase insert/update parcel functions

alexandrutrifu-urzica committed:
VFB-419: updated timestamps and fixed duplicate name in migration

AlexandraIrinaMardale commented on supabase/migrations/20250812124033_add_update_parcel_with_delivery_instructions_function.sql:
```diff
@@ -0,0 +1,41 @@
+        referrer_email      = parcel_record->>'referrer_email',
+        referrer_name       = parcel_record->>'referrer_name',
+        referrer_phone      = parcel_record->>'referrer_phone',
+        voucher_number      = parcel_record->>'voucher_number'
```
Can you double check that these are all the fields that need updating, i.e we haven't added anything else in parcels since this PR was added

AlexandraIrinaMardale commented on supabase/migrations/20250812124058_add_insert_parcel_with_delivery_instructions_function.sql:
```diff
@@ -0,0 +1,52 @@
+    referrer_email,
+    referrer_name,
+    referrer_phone,
+    voucher_number
```
Same here, please

AlexandraIrinaMardale commented on supabase/migrations/20250812124033_add_update_parcel_with_delivery_instructions_function.sql:
Can you rename this to be after the last migration on intern-dev please?
Check with @mihaicutaiulia - she'll add another migration in a bit

AlexandraIrinaMardale reviewed (commented):
😄

alexandrutrifu-urzica committed:
VFB-419: updated migration timestamps
